cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/TriggerJsonCodec.java

```diff
@@ -22,16 +22,27 @@
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
 import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
 
 import java.lang.reflect.Type;
 
 /**
  * Serialization and deserializtion of Triggers as Json.
  *
- * All triggers inherit the {@link Trigger#className} field from the base class. Here we use that
- * during deserialization to determine the actual subclass, and then deserialize based on that.
+ * All triggers must be serialized by passing in the {@link Trigger} class as the type of the object.
```

Comment:
No, but Gson does. That is what I found out during testing. ProgramSchedule has a field of type Trigger. When it serializes the the schedule, it calls the serializer for the Trigger with Trigger.class, and the fields of the subclass don't get serialized, unless the serializer for Trigger explicitly makes a recursive call with the actual subclass as the type. It is exactly contrary to the Javadocs, which states that if you call context.serialize with the same object, it will result in an infinite loop. 
So the previous serialization only worked when serializing Triggers (or subclasses), bit not when serializing objects that contain triggers. I will add a test case for serializing ProgramSchedules (even though that is implicitly tested by the CoreSchedulerServiceTest)